### PR TITLE
Add a preference to always open workspaces in the same window.

### PR DESCRIPTION
**Pull request by epatpol**
Signed-off-by: Patrick-Jeffrey Pollo Guilbert <patrick.pollo.guilbert@ericsson.com>

### CODE REVIEW

**Comment by akosyakov** [comment]
Please elaborate why should it be `always`?

**Comment by epatpol** [comment]
You mean why should the name be `always`? If not I believe some people would prefer not having to open a new tab every time you open a new workspace.

**Comment by akosyakov** [comment]
I understand about changing defaults but i concert about overriding passed arguments, i.e. if someone calls `open({ preserveWindow: false })`, it will be surprising if the window is preserved. What about `window.preserve` option for now with values `on` and `off` which does not override passed arguments? Later if we figure out that it is actually useful to preserve always, we can add another value `always`. I believe there is not a case for it in the codebase so far.

**Comment by epatpol** [comment]
Makes sense, so the option should have priority over the preference as an api, that makes sense.

**Comment by epatpol** [comment]
Also the current change already made sure that passing an option i.e `open({preserveWindow: false})` would override the user preference, so I've changed the naming to not be always but `workspace.preserveWindow`. So option passed as parameter has the highest priority if it exists, then the pref, then the default behavior. Is that better?

File: packages/workspace/src/browser/workspace-service.ts [hunk 1]
```diff
@@ -88,7 +92,10 @@ export class WorkspaceService implements FrontendApplicationContribution {
         const valid = await this.toValidRoot(rootUri);
         if (valid) {
             // The same window has to be preserved too (instead of opening a new one), if the workspace root is not yet available and we are setting it for the first time.
-            const preserveWindow = options ? options.preserveWindow : !(await this.root);
```
Comment (akosyakov) [inline packages/workspace/src/browser/workspace-service.ts]: How about:
```ts
const { preserveWindow } = {
    preserveWindow: this.preferences['workspace.preserveWindow'] || !(await this.root),
    ...options
};
```

It is the pattern which i find useful to define defaults and then override them by arguments if they are provided. After https://github.com/theia-ide/theia/pull/1709#issuecomment-382531492 I've realized that the PR already does what i want but it was not clear for me when i had read the code first time.

Comment (epatpol) [reply packages/workspace/src/browser/workspace-service.ts]: This is much cleaner and easy to understand thanks!

**Comment by akosyakov** [comment]
@epatpol please take care of the build failures, other than that it looks good to me

File: packages/workspace/src/browser/workspace-preferences.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    PreferenceService,
+    PreferenceSchema,
+    PreferenceContribution
+} from '@theia/core/lib/browser/preferences';
+
+export const workspacePreferenceSchema: PreferenceSchema = {
+    "type": "object",
+    "properties": {
+        "workspace.preserveWindow": {
+            "description": "Enable always opening workspaces in current window",
```
Comment (akosyakov) [inline packages/workspace/src/browser/workspace-preferences.ts]: `always` can be omitted?

**Comment by akosyakov** [comment]
@epatpol It seems to work for me if there is already setting.json but if not changes does not seem to be picked up. Try the following:
- start without settings.json
- open a new workspace (should be opened in a new window)
- add settings.json in the original window and set `workspace.preserveWindow` as `true`
- try to open a new workspace from the original window again, it should reuse the original window, but it does not happen for me

**Comment by epatpol** [comment]
So you're adding that settings.json file under the workspace right? I think that's an issue with all workspace preferences right now (I need to test this right now to confirm) that if no workspace settings file was there on startup, adding one won't affect the preferences and that you have to reload for it to work... Maybe this should be fixed also.

**Comment by akosyakov** [comment]
It is a bug then, there should be an event that settings.json is created and preferences should be reread from it. Could you double check and file an issue if it is so?

**Pull request merged**